File: sdk/typescript/consistency.ts
// Typed WriteReceipt / ReadFence helpers for read-after-write consistency.
//
// The broker emits `MutationResponse.write_receipt_json` (field 7) and the
// `x-udb-write-receipt` header on a write, and honors `RequestContext.read_fence_json`
// (field 14) + the `x-udb-read-fence` header on a follow-up read. These helpers
// give callers typed shapes + the load-bearing `source_lsn → min_outbox_lsn`
// mapping so apps never hand-author the Rust-serde-shaped JSON.
//
// The serde field names are pinned by lane 07's machine-derived golden fixture
// (`docs/generated/consistency-golden.json`); these interfaces match it exactly.

import type { CallOptions } from "./generatedClient";

/** A write receipt the broker returns after a mutation. Mirrors the Rust serde
 *  shape `udb::runtime::consistency::WriteReceipt` (all 5 fields always emitted,
 *  no skip_serializing_if). */
export interface WriteReceipt {
  /** Source log sequence number of the committed write (maps to the read
   *  fence's `min_outbox_lsn`). */
  source_lsn: string;
  /** Monotonic outbox sequence for the write. */
  outbox_seq: number;
  /** Projection task ids the write enqueued (empty when none). */
  projection_task_ids: string[];
  /** Catalog/manifest checksum at write time. */
  manifest_checksum: string;
  /** Server commit time, epoch milliseconds. */
  written_at_unix_ms: number;
}

/** A read fence attached to a SPECIFIC follow-up read so it observes the prior
 *  write. Mirrors the Rust serde shape `udb::runtime::consistency::ReadFence`:
 *  `min_outbox_lsn` / `projection_task_ids` are skip-if-empty; `max_wait_ms` is
 *  always serialized. */
export interface ReadFence {
  /** Skip-if-empty: the minimum outbox LSN the read must observe. */
  min_outbox_lsn?: string;
  /** Skip-if-empty: projection task ids that must be applied before the read. */
  projection_task_ids?: string[];
  /** Always serialized: max time the broker waits for the fence, milliseconds. */
  max_wait_ms: number;
}

/** Build a {@link ReadFence} from a {@link WriteReceipt}, applying the
 *  load-bearing cross-type mapping `source_lsn → min_outbox_lsn` and copying the
 *  projection task ids. Empty fields are OMITTED (undefined) so a later
 *  `JSON.stringify` mirrors the Rust `skip_serializing_if`. */
export function readFenceFromReceipt(r: WriteReceipt, maxWaitMs: number): ReadFence {
  return {
    min_outbox_lsn: r.source_lsn ? r.source_lsn : undefined,
    projection_task_ids:
      r.projection_task_ids && r.projection_task_ids.length ? r.projection_task_ids : undefined,
    max_wait_ms: maxWaitMs,
  };
}

/** Parse a receipt JSON string into a typed {@link WriteReceipt}. Returns `null`
 *  on empty/invalid/all-default input so callers never crash on a no-op write
 *  receipt (the broker emits an all-defaults receipt for a write that committed
 *  nothing). */
export function parseWriteReceipt(json: string): WriteReceipt | null {
  if (!json) return null;
  let parsed: unknown;
  try {
    parsed = JSON.parse(json);
  } catch {
    return null;
  }
  if (!parsed || typeof parsed !== "object") return null;
  const obj = parsed as Record<string, unknown>;
  const receipt: WriteReceipt = {
    source_lsn: typeof obj.source_lsn === "string" ? obj.source_lsn : "",
    outbox_seq: typeof obj.outbox_seq === "number" ? obj.outbox_seq : Number(obj.outbox_seq ?? 0),
    projection_task_ids: Array.isArray(obj.projection_task_ids)
      ? (obj.projection_task_ids as string[])
      : [],
    manifest_checksum: typeof obj.manifest_checksum === "string" ? obj.manifest_checksum : "",
    written_at_unix_ms:
      typeof obj.written_at_unix_ms === "number"
        ? obj.written_at_unix_ms
        : Number(obj.written_at_unix_ms ?? 0),
  };
  // An empty (all-default) receipt is a valid no-op signal — still return it so
  // callers can detect it (source_lsn === "" and outbox_seq === 0).
  return receipt;
}

/** Extract a {@link WriteReceipt} from a mutation response body, reading
 *  `write_receipt_json` (MutationResponse field 7). Returns `null` when absent. */
export function receiptFromResponse(resp: any): WriteReceipt | null {
  if (!resp) return null;
  const json = resp.write_receipt_json;
  if (typeof json !== "string") return null;
  return parseWriteReceipt(json);
}

/** Read the durable-idempotency replay flag off a mutation response body,
 *  reading `was_duplicate` (MutationResponse field 3). Returns `true` when the
 *  broker collapsed this call as a REPLAY of a prior write carrying the same
 *  idempotency key (no new row was written), and `false` for a fresh durable
 *  write. Mirrors {@link receiptFromResponse}'s body-reading shape so a caller
 *  that retried (or re-sent) an upsert/delete can tell "my write landed just now"
 *  from "the broker already had this and returned the earlier result".
 *
 *  Accepts either the keepCase wire object (`was_duplicate`) or a camelCase
 *  message (`wasDuplicate`). */
export function wasDuplicate(resp: any): boolean {
  if (!resp || typeof resp !== "object") return false;
  const flag = resp.was_duplicate ?? resp.wasDuplicate;
  return flag === true;
}

// ── Consistency-mode selection (RequestContext.consistency / x-udb-consistency) ─
//
// TS callers had no ergonomic way to declare a read consistency mode (the request
// context only carried `read_fence_json`). These helpers mirror the Python
// selector: they map a friendly mode name to the pinned wire token the broker
// reads at `security.rs` (`x-udb-consistency` header, field 18 `consistency`
// string) and the typed `consistency_mode` enum (field 22). The tokens are PUBLIC
// CONTRACT with `udb::runtime::consistency::ConsistencyMode::as_str` — do not
// rename them.

/** The pinned consistency-mode wire tokens. */
export type ConsistencyModeToken =
  | "strong"
  | "read_your_writes"
  | "bounded_staleness"
  | "replica_bounded"
  | "eventual"
  | "projection_ok"
  | "cache_ok";

/** Friendly, discoverable consistency-mode constants (token-valued, so
 *  `ConsistencyMode.Strong === "strong"`). Pass any of these — or a friendly
 *  alias like `"read-your-writes"` / `"bounded"` — to {@link withConsistency} /
 *  {@link consistencyContext}. */
export const ConsistencyMode = {
  /** Read the primary, no replica fallback (linearizable). */
  Strong: "strong",
  /** Read your own writes — primary or a replica fenced past your write LSN. */
  ReadYourWrites: "read_your_writes",
  /** Any replica caught up within `max_replica_lag_ms` (bounded staleness). */
  BoundedStaleness: "bounded_staleness",
  /** A physical read replica within a lag budget, failing over to the primary. */
  ReplicaBounded: "replica_bounded",
  /** Any healthy replica/projection; no fence, cheapest. */
  Eventual: "eventual",
  /** Projection-backed targets (Mongo/Qdrant/ClickHouse) acceptable. */
  ProjectionOk: "projection_ok",
  /** Redis cache hits acceptable (cache verifies manifest checksum). */
  CacheOk: "cache_ok",
} as const;

// Friendly aliases → canonical token. Mirrors the broker's lenient `parse`
// (consistency.rs) plus the short spec names ("bounded" → bounded_staleness).
const CONSISTENCY_ALIASES: Record<string, ConsistencyModeToken> = {
  strong: "strong",
  linearizable: "strong",
  primary: "strong",
  read_your_writes: "read_your_writes",
  "read-your-writes": "read_your_writes",
  ryw: "read_your_writes",
  bounded: "bounded_staleness",
  bounded_staleness: "bounded_staleness",
  "bounded-staleness": "bounded_staleness",
  replica_bounded: "replica_bounded",
  "replica-bounded": "replica_bounded",
  eventual: "eventual",
  eventual_consistency: "eventual",
  projection_ok: "projection_ok",
  "projection-ok": "projection_ok",
  cache_ok: "cache_ok",
  "cache-ok": "cache_ok",
};

// Canonical token → proto enum NAME (context.proto field 22). Under the keepCase
// / `enums: String` loader the wire carries the enum NAME string, not the number.
const CONSISTENCY_MODE_ENUM_NAME: Record<ConsistencyModeToken, string> = {
  strong: "CONSISTENCY_MODE_STRONG",
  read_your_writes: "CONSISTENCY_MODE_READ_YOUR_WRITES",
  bounded_staleness: "CONSISTENCY_MODE_BOUNDED_STALENESS",
  replica_bounded: "CONSISTENCY_MODE_REPLICA_BOUNDED",
  eventual: "CONSISTENCY_MODE_EVENTUAL",
  projection_ok: "CONSISTENCY_MODE_PROJECTION_OK",
  cache_ok: "CONSISTENCY_MODE_CACHE_OK",
};

/** Resolve a mode name/alias to its canonical wire token. Throws on an unknown
 *  mode so a typo fails loudly rather than silently defaulting to strong. */
export function consistencyToken(mode: ConsistencyModeToken | string): ConsistencyModeToken {
  const token = CONSISTENCY_ALIASES[String(mode).trim().toLowerCase()];
  if (!token) {
    throw new Error(
      `udb: unknown consistency mode "${mode}" (expected one of ` +
        `strong | read_your_writes | bounded_staleness | replica_bounded | eventual | projection_ok | cache_ok)`,
    );
  }
  return token;
}

/** Returns a {@link CallOptions} that declares the read consistency `mode` on
 *  exactly this one call via the `x-udb-consistency` header (the header the
 *  broker reads at `security.rs`; it WINS over the body `consistency` field when
 *  both are present). Mirrors the Python `x-udb-consistency` selector. */
export function withConsistency(mode: ConsistencyModeToken | string): CallOptions {
  return { headers: { "x-udb-consistency": consistencyToken(mode) } };
}

/** For body-context callers: returns a partial `RequestContext` setting both the
 *  legacy `consistency` string (context.proto field 18) and the typed
 *  `consistency_mode` enum name (field 22). Spread into the request `context`.
 *  The `x-udb-consistency` header still wins server-side when both are supplied. */
export function consistencyContext(
  mode: ConsistencyModeToken | string,
): { consistency: ConsistencyModeToken; consistency_mode: string } {
  const token = consistencyToken(mode);
  return { consistency: token, consistency_mode: CONSISTENCY_MODE_ENUM_NAME[token] };
}

/** Returns a {@link CallOptions} that attaches `fence` to exactly the one
 *  follow-up read via the `x-udb-read-fence` header (the broker reads it at
 *  security.rs / service/mod.rs). The fence rides this single call only — it is
 *  NEVER stored on shared project metadata. */
export function withReadFence(fence: ReadFence): CallOptions {
  return { headers: { "x-udb-read-fence": JSON.stringify(fence) } };
}

/** For body-context callers: returns a partial `RequestContext` setting
 *  `read_fence_json` (context.proto field 14). Spread into the request `context`. */
export function readFenceContext(fence: ReadFence): { read_fence_json: string } {
  return { read_fence_json: JSON.stringify(fence) };
}

/** One-shot helper (naming contract alias): convert a {@link WriteReceipt} into a
 *  {@link ReadFence} and attach it to the next read via `x-udb-read-fence`.
 *  Composes {@link readFenceFromReceipt} + {@link withReadFence} — no round trip. */
export function withReadFenceFromReceipt(r: WriteReceipt, maxWaitMs: number): CallOptions {
  return withReadFence(readFenceFromReceipt(r, maxWaitMs));
}

/** Canonical receipt accessor (naming contract): attach the read fence derived
 *  from a write `receipt` to exactly the next read. Alias of
 *  {@link withReadFenceFromReceipt}; the spec name is `metadata.afterWrite(receipt)`
 *  (exposed as `UdbProject.metadata.afterWrite`). The fence rides only this one
 *  CallOptions — never shared project metadata. */
export function afterWrite(receipt: WriteReceipt, maxWaitMs = 5_000): CallOptions {
  return withReadFenceFromReceipt(receipt, maxWaitMs);
}

/** Grouped receipt/fence accessors. `UdbProject.metadata` is an instance of this
 *  shape so callers can write `udb.metadata.afterWrite(receipt)` per the naming
 *  contract. Standalone functions of the same name remain exported directly. */
export const consistencyMetadata = {
  /** Alias of {@link afterWrite} — attach a receipt-derived read fence to the next read. */
  afterWrite,
  /** Alias of {@link withReadFenceFromReceipt}. */
  withReadFenceFromReceipt,
  /** Alias of {@link withReadFence}. */
  withReadFence,
  /** Alias of {@link readFenceFromReceipt}. */
  readFenceFromReceipt,
  /** Alias of {@link receiptFromResponse}. */
  receiptFromResponse,
  /** Alias of {@link wasDuplicate} — durable-idempotency replay flag off a mutation. */
  wasDuplicate,
  /** Alias of {@link withConsistency} — attach a consistency mode to the next call. */
  withConsistency,
  /** Alias of {@link consistencyContext} — body-context consistency-mode fields. */
  consistencyContext,
  /** Alias of {@link consistencyToken} — resolve a mode name to its wire token. */
  consistencyToken,
  /** The friendly consistency-mode constants ({@link ConsistencyMode}). */
  ConsistencyMode,
};

File: sdk/go/udbclient/consistency.go
package udbclient

import (
	"encoding/json"

	entityv1 "github.com/fahara02/udb/sdk/go/gen/udb/entity/v1"
)

// ── Typed write-receipt / read-fence helpers (chapter 08.1) ──────────────────
//
// The broker emits MutationResponse.write_receipt_json (and an x-udb-write-receipt
// header) per write, and RequestContext carries read_fence_json for read-your-writes.
// These types mirror the Rust serde shapes 1:1 so apps never hand-author
// Rust-shaped JSON. The load-bearing cross-type mapping is
// WriteReceipt.SourceLsn -> ReadFence.MinOutboxLsn (see ReadFenceFromReceipt).
//
// Both shapes are pinned byte-for-byte to the committed, machine-derived golden
// fixture docs/generated/consistency-golden.json (lane 07) by
// consistency_test.go — do NOT drift the JSON tags from that fixture.

// WriteReceipt mirrors the Rust `WriteReceipt` struct (src/runtime/consistency.rs).
// The Rust struct serializes ALL five fields unconditionally (no
// skip_serializing_if), so none of the JSON tags carry `omitempty`.
type WriteReceipt struct {
	SourceLsn         string   `json:"source_lsn"`
	OutboxSeq         uint64   `json:"outbox_seq"`
	ProjectionTaskIds []string `json:"projection_task_ids"`
	ManifestChecksum  string   `json:"manifest_checksum"`
	WrittenAtUnixMs   int64    `json:"written_at_unix_ms"`
}

// ParseWriteReceipt decodes write_receipt_json bytes into a WriteReceipt.
// Empty input yields a zero receipt and a nil error (no write produced a receipt).
func ParseWriteReceipt(b []byte) (WriteReceipt, error) {
	if len(b) == 0 {
		return WriteReceipt{}, nil
	}
	var r WriteReceipt
	if err := json.Unmarshal(b, &r); err != nil {
		return WriteReceipt{}, err
	}
	return r, nil
}

// IsEmpty reports whether the receipt carries no write information.
func (r WriteReceipt) IsEmpty() bool {
	return r.SourceLsn == "" && r.OutboxSeq == 0 &&
		len(r.ProjectionTaskIds) == 0 && r.ManifestChecksum == "" && r.WrittenAtUnixMs == 0
}

// ReadFence mirrors the Rust `ReadFence` struct (src/runtime/consistency.rs).
// In Rust, min_outbox_lsn and projection_task_ids carry skip_serializing_if
// (empty omitted); max_wait_ms is `#[serde(default)]` only (always emitted), so
// it has NO omitempty here.
type ReadFence struct {
	MinOutboxLsn      string   `json:"min_outbox_lsn,omitempty"`
	ProjectionTaskIds []string `json:"projection_task_ids,omitempty"`
	MaxWaitMs         uint64   `json:"max_wait_ms"`
}

// IsEmpty reports whether the fence carries no positional constraint (only a
// wait budget, or nothing at all).
func (f ReadFence) IsEmpty() bool {
	return f.MinOutboxLsn == "" && len(f.ProjectionTaskIds) == 0
}

// ReadFenceFromReceipt builds the fence a follow-up read attaches to wait for its
// own write to be visible. It maps the receipt's SourceLsn -> MinOutboxLsn (the
// load-bearing cross-type field mapping from Rust `ReadFence::from_receipt`) and
// copies the projection task ids; it does NOT carry over outbox_seq or
// manifest_checksum.
func ReadFenceFromReceipt(r WriteReceipt, maxWaitMs uint64) ReadFence {
	return ReadFence{
		MinOutboxLsn:      r.SourceLsn,
		ProjectionTaskIds: append([]string(nil), r.ProjectionTaskIds...),
		MaxWaitMs:         maxWaitMs,
	}
}

// ReceiptFromMutation captures the WriteReceipt from a MutationResponse body
// field (the primary capture path; the x-udb-write-receipt header is the
// forward/embedded fallback). An empty body field yields a zero receipt and a nil
// error.
func ReceiptFromMutation(m *entityv1.MutationResponse) (WriteReceipt, error) {
	if m == nil {
		return WriteReceipt{}, nil
	}
	return ParseWriteReceipt([]byte(m.GetWriteReceiptJson()))
}

// WasDuplicate reports whether the broker collapsed this mutation onto a prior
// write via durable idempotency (a replay of the same idempotency key) instead
// of applying a fresh write. It mirrors MutationResponse.was_duplicate and lets a
// caller of the raw Upsert/Delete path distinguish an idempotency replay from a
// fresh mutation (Entity.Upsert also surfaces this on UpsertResult.WasDuplicate).
// A nil response yields false.
func WasDuplicate(m *entityv1.MutationResponse) bool {
	return m.GetWasDuplicate()
}

// ── Consistency-mode selection (chapter 08.1) ────────────────────────────────
//
// ConsistencyMode is an ergonomic selector for the read/write consistency a
// single request wants, mirroring Python's Metadata consistency knob. Its string
// value is the PINNED wire token the broker parses (src/runtime/consistency.rs
// ConsistencyMode::parse) — from either the x-udb-consistency header or the
// typed RequestContext.consistency_mode enum (proto field 22) it maps 1:1 onto.
// Changing a token breaks every client, so these stay byte-for-byte pinned.
type ConsistencyMode string

const (
	// ConsistencyDefault leaves the mode unset (broker default = strong).
	ConsistencyDefault ConsistencyMode = ""
	// ConsistencyStrong reads the primary / linearizable copy.
	ConsistencyStrong ConsistencyMode = "strong"
	// ConsistencyReadYourWrites guarantees a session observes its own writes.
	ConsistencyReadYourWrites ConsistencyMode = "read_your_writes"
	// ConsistencyBoundedStaleness allows a replica read within a lag bound.
	ConsistencyBoundedStaleness ConsistencyMode = "bounded_staleness"
	// ConsistencyReplicaBounded prefers a replica within a bound.
	ConsistencyReplicaBounded ConsistencyMode = "replica_bounded"
	// ConsistencyEventual allows any replica (fastest, weakest).
	ConsistencyEventual ConsistencyMode = "eventual"
	// ConsistencyProjectionOk permits serving from an async projection.
	ConsistencyProjectionOk ConsistencyMode = "projection_ok"
	// ConsistencyCacheOk permits serving from a cache.
	ConsistencyCacheOk ConsistencyMode = "cache_ok"
)

// proto maps the wire token onto the entity/v1.ConsistencyMode enum the broker
// reads from RequestContext.consistency_mode. Unknown/empty maps to UNSPECIFIED.
func (m ConsistencyMode) proto() entityv1.ConsistencyMode {
	switch m {
	case ConsistencyStrong:
		return entityv1.ConsistencyMode_CONSISTENCY_MODE_STRONG
	case ConsistencyReadYourWrites:
		return entityv1.ConsistencyMode_CONSISTENCY_MODE_READ_YOUR_WRITES
	case ConsistencyBoundedStaleness:
		return entityv1.ConsistencyMode_CONSISTENCY_MODE_BOUNDED_STALENESS
	case ConsistencyReplicaBounded:
		return entityv1.ConsistencyMode_CONSISTENCY_MODE_REPLICA_BOUNDED
	case ConsistencyEventual:
		return entityv1.ConsistencyMode_CONSISTENCY_MODE_EVENTUAL
	case ConsistencyProjectionOk:
		return entityv1.ConsistencyMode_CONSISTENCY_MODE_PROJECTION_OK
	case ConsistencyCacheOk:
		return entityv1.ConsistencyMode_CONSISTENCY_MODE_CACHE_OK
	default:
		return entityv1.ConsistencyMode_CONSISTENCY_MODE_UNSPECIFIED
	}
}

// Apply stamps this consistency mode onto a single, caller-supplied per-read
// RequestContext (the typed consistency_mode enum the broker honors). It NEVER
// touches shared Udb/facade metadata — the choice rides only this one request. An
// empty mode leaves rc untouched.
func (m ConsistencyMode) Apply(rc *entityv1.RequestContext) {
	if rc == nil || m == ConsistencyDefault {
		return
	}
	rc.ConsistencyMode = m.proto()
}

// Header returns the (key, value) metadata pair for the x-udb-consistency header
// form of this mode, for callers stamping consistency via request headers rather
// than the RequestContext body. An empty mode yields an empty value.
func (m ConsistencyMode) Header() (string, string) {
	return "x-udb-consistency", string(m)
}

// ── Facade-mounted metadata accessor (naming contract R2.0) ──────────────────

// MetadataAccessor is the metadata surface mounted on the project facade via
// Udb.Metadata(), matching the cross-language udb.metadata.* shape (TS
// udb.metadata.afterWrite, Python udb.metadata). It groups the per-request
// RequestContext stamping helpers so callers reach them off the project object
// instead of the package-level functions.
type MetadataAccessor struct{ u *Udb }

// Metadata returns the metadata accessor mounted on the project facade so
// udb.Metadata().AfterWrite(rc, receipt, maxWaitMs) mirrors TS
// udb.metadata.afterWrite(receipt).
func (u *Udb) Metadata() *MetadataAccessor { return &MetadataAccessor{u: u} }

// AfterWrite stamps a read fence derived from a write receipt onto a single
// per-read RequestContext so the follow-up read observes its own write. It
// delegates to the package-level AfterWrite and never touches shared facade
// metadata — the fence rides only this one request.
func (a *MetadataAccessor) AfterWrite(rc *entityv1.RequestContext, r WriteReceipt, maxWaitMs uint64) {
	AfterWrite(rc, r, maxWaitMs)
}

// Consistency stamps a consistency mode onto a single per-read RequestContext
// (delegates to ConsistencyMode.Apply).
func (a *MetadataAccessor) Consistency(rc *entityv1.RequestContext, mode ConsistencyMode) {
	mode.Apply(rc)
}

// AfterWrite stamps a read fence derived from a write receipt onto a single,
// caller-supplied per-read RequestContext so the follow-up read observes its own
// write. It NEVER touches shared Udb/facade metadata — the fence rides only this
// one request (guardrail: no leaking a fence into unrelated reads). An empty
// receipt leaves rc untouched.
func AfterWrite(rc *entityv1.RequestContext, r WriteReceipt, maxWaitMs uint64) {
	if rc == nil || r.IsEmpty() {
		return
	}
	fence := ReadFenceFromReceipt(r, maxWaitMs)
	b, err := json.Marshal(fence)
	if err != nil {
		return
	}
	rc.ReadFenceJson = string(b)
}
